### PR TITLE
Support for 12-hour clock (AM/PM) for TimePicker component

### LINKED ISSUE #1409: [DateTime] Support 12hr (AM/PM) times as well
### Feature Request

Can the [TimePicker](http://blueprintjs.com/docs/#datetime/timepicker) be enhanced to allow for either 12hr or 24hr times? 

Thanks!

// Copied from #1810 

> datetime v1.24.0
> 
> it would be nice if it was more clear in the ui that the timepicker is a 24hr format. The way it is right now, it's possible users enter 1:00 meaning 1:00PM and never discover they are wrong
> 
> it would be nice if there was an option to be in AM/PM format

### PR DESCRIPTION
**Pull request by colinbr96**
#### Checklist

- [x] [Enable CircleCI for your fork](https://circleci.com/add-projects)
- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

Adding an additional prop ~~`hourFormat`~~ `useAmPm` to the TimePicker component to support a 12-hour clock and AM/PM switcher. ~~`hourFormat` can be set to either of the two strings `"12hour"` or `"24hour"` to toggle the AM/PM switcher and type of hour unit being used.~~ `useAmPm` is a boolean which defaults to `false` (using the 24-hour clock by default to maintain backwards compatibility) and can be set to `true` to set the hour range to 12-hours and display an AM/PM dropdown.

I also added a switch to the TimePicker documentation page to set the example component's `hourFormat`.

#### Video Demo

https://media.giphy.com/media/39m89EvdXoQFbCPKYk/giphy.gif

### CODE REVIEW

**Comment by colinbr96** [comment]
FYI, I am currently in progress on adding test cases and linting so that CI will accept this branch.

**Comment by giladgray** [comment]
@colinbr96 amazing work! Thanks in advance for tests.

API-wise, I'm not a fan of the two strings
An enum is better, or even a boolean as there are only two options, but I'm not settled on a name. Perhaps something like `twelveHour` or `showAmPm`?

Finally, would you kindly configure your circle build so the preview comment will post correctly? There's a wiki page with instructions for the environment variable.

**Comment by colinbr96** [comment]
<h3>Changed API to useAmPm prop</h3>

Preview: <strong><a href='https://13-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://13-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://13-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

**Comment by colinbr96** [comment]
I have updated the API to use a boolean instead. Now the TimePicker prop will be called `useAmPm`. By default, it will be set to false (so that the 24-hour clock is used initially).

**Comment by colinbr96** [comment]
<h3>Added tests for remaining dateUtils functions</h3>

Preview: <strong><a href='https://19-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://19-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://19-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

File: packages/datetime/src/_timepicker.scss [hunk 1]
```diff
@@ -73,6 +74,10 @@ $timepicker-control-width: $pt-grid-size * 3.3 !default;
     }
   }
 
+  .pt-timepicker-ampm-select {
+    margin-left: 2px;
```
Comment (giladgray) [inline packages/datetime/src/_timepicker.scss]: `$pt-grid-size / 2` please: typical small spacing.

File: packages/datetime/src/common/dateUtils.ts [hunk 2]
```diff
@@ -205,3 +205,46 @@ export function getDateNextMonth(date: Date): Date {
...
+        throw new Error(`hour must be between [0,23] inclusive: got ${hour}`);
+    }
+    return hour >= 12;
+}
+
+export function get12HourFrom24Hour(hour: number): number {
+    if (hour < 0 || hour > 23) {
+        throw new Error(`hour must be between [0,23] inclusive: got ${hour}`);
+    }
+    hour = hour % 12;
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: 🙅‍♂️ please refrain from modifying arguments. create new `const` variables.

File: packages/datetime/src/common/dateUtils.ts [hunk 3]
```diff
@@ -205,3 +205,46 @@ export function getDateNextMonth(date: Date): Date {
...
+    if (hour === 0) {
+        hour = 12;
+    }
+    return hour;
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: we love ternary statements here: `return hour === 0 ? 12 : hour;`

File: packages/datetime/src/common/dateUtils.ts [hunk 4]
```diff
@@ -205,3 +205,46 @@ export function getDateNextMonth(date: Date): Date {
...
+}
+
+export function get24HourFrom12Hour(hour: number, isPm: boolean): number {
+    if (hour < 1 || hour > 12) {
+        throw new Error(`hour must be between [0,12] inclusive: got ${hour}`);
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: `< 1` and `[0,12]` don't match up

File: packages/datetime/src/common/dateUtils.ts [hunk 5]
```diff
@@ -205,3 +205,46 @@ export function getDateNextMonth(date: Date): Date {
...
+    }
+    if (isPm) {
+        hour = hour + 12;
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: avoid modifying arguments. new variables please.

File: packages/datetime/src/common/dateUtils.ts [hunk 6]
```diff
@@ -205,3 +205,46 @@ export function getDateNextMonth(date: Date): Date {
...
+        if (hour === 24) {
+            hour = 12;
+        }
+    } else if (hour === 12) {
+        hour = 0;
+    }
+    return hour;
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: how about:

```ts
const newHour = hour === 12 ? 0 : hour;
return isPm ? newHour + 12 : newHour;
```

File: packages/datetime/src/timePicker.tsx [hunk 7]
```diff
@@ -207,6 +230,41 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
         );
     }
 
+    private maybeRenderAmPm() {
+        if (!this.props.useAmPm) {
+            return null;
+        }
+
+        const onChange = (amPm: string) => {
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: pull this to a member function: `private handleAmPmChange`.
currently, it gets redefined every render and breaks any pureness as it's always a new instance.

File: packages/datetime/test/common/dateUtilsTests.tsx [hunk 8]
```diff
@@ -221,4 +221,119 @@ describe("dateUtils", () => {
             assertTimeIs(time, 11, 20);
         });
     });
+
+    describe("convert24HourMeridiem", () => {
+        it("returns given hour, if hour is AM and toAm", () => {
+            for (let hour = 0; hour <= 11; hour++) {
```
Comment (giladgray) [inline packages/datetime/test/common/dateUtilsTests.tsx]: for all the tests you've added here, please just test one or two explicit values. 
we gain nothing by iterating through math, but we lose much in readability.

try this:
```ts
expect(DateUtils.convert24HourMeridiem(4, true)).to.equal(4)
expect(DateUtils.convert24HourMeridiem(16, true)).to.equal(4) // < edge cases! tests!
```

File: packages/datetime/test/common/dateUtilsTests.tsx [hunk 9]
```diff
@@ -221,4 +221,119 @@ describe("dateUtils", () => {
...
+            for (const hour of amHours) {
+                expect(DateUtils.getIsPmFrom24Hour(hour)).to.equal(false);
+            }
+        });
+
+        it("throws an error only for invalid hours", () => {
+            expect(() => DateUtils.getIsPmFrom24Hour(-1)).to.throw();
+            expect(() => DateUtils.getIsPmFrom24Hour(24)).to.throw();
+            expect(() => DateUtils.getIsPmFrom24Hour(0)).to.not.throw();
+            expect(() => DateUtils.getIsPmFrom24Hour(23)).to.not.throw();
```
Comment (giladgray) [inline packages/datetime/test/common/dateUtilsTests.tsx]: nice, _that's_ how you test a few demonstrative values 👍

File: packages/datetime/src/timePicker.tsx [hunk 10]
```diff
@@ -375,3 +439,8 @@ function handleKeyEvent(e: React.KeyboardEvent<HTMLInputElement>, actions: IKeyE
         }
     }
 }
+
+/** Event handler that exposes the target element's value as a string. */
+function handleStringChange(handler: (value: string) => void) {
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: please remove this. it's my favorite util function, too, but it does not belong here. 

🤔 we should move them from docs-theme to core utils, in a future PR. #helpwanted 😄

File: packages/datetime/src/timePicker.tsx [hunk 11]
```diff
@@ -207,6 +230,41 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
...
+                    value={this.state.isPm ? "pm" : "am"}
+                    onChange={handleStringChange(onChange)}
+                    disabled={this.props.disabled}
+                >
+                    <option key={0} value={"am"}>
+                        AM
+                    </option>
+                    <option key={1} value={"pm"}>
+                        PM
+                    </option>
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: `key` is only necessary when rendering an array. refactor these to: `<option value="pm">PM</option>`

File: packages/datetime/src/timePicker.tsx [hunk 12]
```diff
@@ -163,15 +179,22 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
         const didMinTimeChange = nextProps.minTime !== this.props.minTime;
         const didMaxTimeChange = nextProps.maxTime !== this.props.maxTime;
         const didBoundsChange = didMinTimeChange || didMaxTimeChange;
+        const didUseAmPmChange = nextProps.useAmPm !== this.props.useAmPm;
+
+        let value = this.state.value;
+        let useAmPm = this.props.useAmPm || TimePicker.defaultProps.useAmPm;
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: no need for the `|| defaultProps` here. React already does that for us 😄

File: packages/datetime/src/timePicker.tsx [hunk 13]
```diff
@@ -105,25 +112,33 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
...
+        useAmPm: false,
     };
 
     public static displayName = "Blueprint2.TimePicker";
 
     public constructor(props?: ITimePickerProps, context?: any) {
         super(props, context);
 
+        let value = props.minTime;
+        let useAmPm = TimePicker.defaultProps.useAmPm;
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: this is not necessary. `props.useAmPm` will already be set to this. you can remove this variable entirely.

File: packages/datetime/src/timePicker.tsx [hunk 14]
```diff
@@ -105,25 +112,33 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
...
-            this.state = this.getFullStateFromValue(props.defaultValue);
-        } else {
-            this.state = this.getFullStateFromValue(props.minTime);
+            value = props.defaultValue;
+        }
+        if (props.useAmPm != null) {
+            useAmPm = props.useAmPm;
         }
+
+        this.state = this.getFullStateFromValue(value, useAmPm);
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: just pass `props.useAmPm` along here

**Comment by colinbr96** [comment]
Thanks for the code review. I got the requested changes completed. Hopefully it's better this _time_! 🕰️

**Comment by colinbr96** [comment]
<h3>Remove explicit references to defaultProps for useAmPm</h3>

Preview: <strong><a href='https://25-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://25-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://25-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

File: packages/datetime/test/common/dateUtilsTests.tsx [hunk 15]
```diff
@@ -265,21 +265,19 @@ describe("dateUtils", () => {
...
         it("returns true, if hour (in 24 range) is PM", () => {
-            const pmHours = [];
-            for (let i = 12; i <= 23; i++) {
-                pmHours.push(i);
-            }
-            for (const hour of pmHours) {
+            expect(DateUtils.getIsPmFrom24Hour(12)).to.equal(true);
+            expect(DateUtils.getIsPmFrom24Hour(23)).to.equal(true);
+
+            for (let hour = 12; hour <= 23; hour++) {
```
Comment (giladgray) [inline packages/datetime/test/common/dateUtilsTests.tsx]: wait wait the point of my previous comment was to _replace_ all `for` loops with a handful of explicit tests. you added explicit tests but didn't _remove_ the loops.

no need to test 12 and 23 _and then_ test 12 _through_ 23.

(this comment applies to every test in this file. i want to see zero `for` loops when we're done.)

File: packages/datetime/src/timePicker.tsx [hunk 16]
```diff
@@ -234,32 +229,15 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
...
                     disabled={this.props.disabled}
                 >
-                    <option key={0} value={"am"}>
-                        AM
-                    </option>
-                    <option key={1} value={"pm"}>
-                        PM
-                    </option>
+                    <option value={"am"}>AM</option>
+                    <option value={"pm"}>PM</option>
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: braces are not necessary for string literals. `value="pm"`

File: packages/datetime/src/timePicker.tsx [hunk 17]
```diff
@@ -163,15 +174,22 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
...
-            const timeInRange = DateUtils.getTimeInRange(this.state.value, nextProps.minTime, nextProps.maxTime);
-            this.setState(this.getFullStateFromValue(timeInRange));
+            value = DateUtils.getTimeInRange(this.state.value, nextProps.minTime, nextProps.maxTime);
         }
-
         if (nextProps.value != null && !DateUtils.areSameTime(nextProps.value, this.props.value)) {
-            this.setState(this.getFullStateFromValue(nextProps.value));
+            value = nextProps.value;
+        }
+        if (didUseAmPmChange) {
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: i think this check is useless and it is equivalent to just passing through `nextProps.useAmPm`. if props are unchanged then next is the same as `this.props`. please remove (just like we did on line 130).

File: packages/datetime/src/timePicker.tsx [hunk 18]
```diff
@@ -64,6 +64,12 @@ export interface ITimePickerProps extends IProps {
      */
     showArrowButtons?: boolean;
 
+    /**
+     * Whether to choose from a 12 hour format with an AM/PM dropdown.
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: "Whether to [use a] 12-hour format with an AM/PM dropdown."

File: packages/datetime/test/common/dateUtilsTests.tsx (hunk 8 above)
Comment (giladgray) [reply packages/datetime/test/common/dateUtilsTests.tsx]: this comment still stands. please address it in all tests.

File: packages/datetime/test/common/dateUtilsTests.tsx [hunk 19]
```diff
@@ -221,4 +221,124 @@ describe("dateUtils", () => {
...
+    });
+
+    describe("getIsPmFrom24Hour", () => {
+        it("returns true, if hour (in 24 range) is PM", () => {
+            expect(DateUtils.getIsPmFrom24Hour(12)).to.equal(true);
+            expect(DateUtils.getIsPmFrom24Hour(23)).to.equal(true);
+
+            for (let hour = 12; hour <= 23; hour++) {
+                expect(DateUtils.getIsPmFrom24Hour(hour)).to.equal(true);
+            }
```
Comment (giladgray) [inline packages/datetime/test/common/dateUtilsTests.tsx]: so you added some explicit values, now it's time to remove the `for` loop.
(this applies to several tests here)

File: packages/datetime/test/common/dateUtilsTests.tsx (hunk 15 above)
Comment (colinbr96) [reply packages/datetime/test/common/dateUtilsTests.tsx]: does it make sense to explicitly test for all values from 12 to 23? Normally I would just iterate this
**e.g.**
`expect(DateUtils.getIsPmFrom24Hour(12)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(13)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(14)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(15)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(16)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(17)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(18)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(19)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(20)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(21)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(22)).to.equal(true);`

`expect(DateUtils.getIsPmFrom24Hour(23)).to.equal(true);`

**Comment by colinbr96** [comment]
<h3>Fix misc code review requests</h3>

Preview: <strong><a href='https://31-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/docs-app/dist/index.html' target='_blank'>documentation</a> | <a href='https://31-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/landing-app/dist/index.html' target='_blank'>landing</a> | <a href='https://31-121902933-gh.circle-artifacts.com/0/home/circleci/project/packages/table-dev-app/dist/index.html' target='_blank'>table</a></strong>

**Comment by colinbr96** [comment]
@giladgray For some reason this pull request still says "changes requested" however I have already finished making the requested changes 3 days ago. Not sure what I need to do next.
